### PR TITLE
Add support for DynamodbEvent.DynamodbStreamRecord serialization

### PR DESCRIPTION
**Pull request by mo7ty**
*Issue #452:*
Failed to deserialize DynamodbEvent.DynamodbStreamRecord#dynamodb

*Description of changes:*
Add support for DynamodbEvent.DynamodbStreamRecord serialization

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        },
+        "OldImage": {
+            "Message": {
+                "S": "New item!"
+            },
+            "Id": {
+                "N": "101"
+            }
+        },
+        "ApproximateCreationDateTime": 1428537600,
```
Comment (smirnoal) [inline aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json:29]: I think this is represented as a floating point number in this event, i.e.`1428537600.0`

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json:29]: The already existing [record with `eventName=MODIFY` in `dynamo_event.json`](https://github.com/aws/aws-lambda-java-libs/blob/02e3cdc3caa6a4aee2fb4c3c221191994709e080/aws-lambda-java-tests/src/test/resources/ddb/dynamo_event.json#L34-L68) was reused.
According to [StreamRecord.ApproximateCreationDateTime - Amazon DynamoDB](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_streams_StreamRecord.html#:~:text=ApproximateCreationDateTime) documentation, a `long` number should be considered:

> The approximate date and time when the stream record was created, in [UNIX epoch time](http://www.epochconverter.com/) format and rounded down to the closest second.
> Type: Timestamp

Also, no decimals are considered in the (_not edited_) [corresponding test assertion](https://github.com/aws/aws-lambda-java-libs/pull/455/files/02e3cdc3caa6a4aee2fb4c3c221191994709e080#diff-46ce264a2374bf6be6602d1f35ac3e89719b50bc7e11ed5acefc34e791a8eaffR185) using:
[java.util.Date.from](https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#from-java.time.Instant-)([java.time.Instant.ofEpochSecond(long)](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#ofEpochSecond-long-)).

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json:29]: Hi @smirnoal,

In AWS you might see a dotted number representation for `StreamRecord.ApproximateCreationDateTime`, but is its scientific (number in exponential) notation - observed sample:

```
ApproximateCreationDateTime=1.697900578E9
```

The `long` number representation for [UNIX epoch time](http://www.epochconverter.com/) `1697900578` - ISO8601: `2023-10-21T15:02:58Z`.

Comment (smirnoal) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json:29]: I recently saw event like below from PROD
```
{
    'Records': [{
        'eventID': '5f342e95ccef0cf90ff71be3fc9c96e5',
        'eventName': 'INSERT',
        'eventVersion': '1.1',
        'eventSource': 'aws:dynamodb',
        'awsRegion': 'us-east-1',
        'dynamodb': {
            'ApproximateCreationDateTime': 1697219385.0,
            'Keys': {
                'date': {
                    'S': '20230101'
                },
                'playerId': {
                    'S': 'abcd'
                }
            },
            'NewImage': {
                'date': {
                    'S': '20230101'
                },
                'playerId': {
                    'S': 'abcd'
                }
            },
            'SequenceNumber': '3131009900000000046478534043',
            'SizeBytes': 48,
            'StreamViewType': 'NEW_AND_OLD_IMAGES'
        },
        'eventSourceARN': 'arn:aws:dynamodb:us-east-1:<accountid>:table/GameEvents/stream/2021-12-24T20:58:44.015'
    }]
}
```

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: Updated in commit 232497ac.

Comment (vaibhav170) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: We have seen epoch format for  ApproximateCreationDateTime attribute in dynamodb event
"ApproximateCreationDateTime": 1698752983,

And dynamodb event via kinesis is in millisecond precision  
"ApproximateCreationDateTime":1698658039410,

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: Hi @vaibhav170,

Thanks for the feedback.

Do you think it worth adding different tests for each and every possible and already observed `ApproximateCreationDateTime` value when AWS states that is [rounded down to the closest second](https://github.com/aws/aws-lambda-java-libs/pull/455#discussion_r1372401195)?

Comment (vaibhav170) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: i think second level precision test is fine for dynamodb stream event to align with documentation

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: Updated in commit f92b0bc5.

Comment (mo7ty) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: Hi @smirnoal,

Sorry to ask, but what is required for this PR to be approved&merged?

Thanks

**Comment by pedrogcsb** [comment]
This is usefull for a project I'm currently working on. Can this be released?

File: aws-lambda-java-tests/src/test/resources/ddb/dynamo_event.json [hunk 2]
```diff
@@ -59,7 +59,7 @@
             "N": "101"
           }
         },
-        "ApproximateCreationDateTime": 1428537600,
+        "ApproximateCreationDateTime": 1.635734407123456789E9,
```
Comment (jeromevdl) [inline aws-lambda-java-tests/src/test/resources/ddb/dynamo_event.json:62]: are you sure about this format ?

File: aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json (hunk 1 above)
Comment (smirnoal) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: hey sorry was a bit busy with java21 launch. Will get to this next week.

**Review by smirnoal — CHANGES_REQUESTED** [review]

File: aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json (hunk 1 above)
Comment (smirnoal) [reply aws-lambda-java-tests/src/test/resources/ddb/dynamo_ddb_stream_record.json]: I think you will need to rebase it before merging: "This branch is out-of-date with the base branch"

File: aws-lambda-java-tests/src/main/java/com/amazonaws/services/lambda/runtime/tests/EventLoader.java [hunk 3]
```diff
@@ -65,6 +65,10 @@ public static DynamodbEvent loadDynamoDbEvent(String filename) {
         return loadEvent(filename, DynamodbEvent.class);
     }
 
+    public static DynamodbEvent.DynamodbStreamRecord loadDynamodbStreamRecord(String filename) {
```
[the remark this answers is not in the document]
Comment (mo7ty) [reply aws-lambda-java-tests/src/main/java/com/amazonaws/services/lambda/runtime/tests/EventLoader.java]: Updated in commit ffafb30d.

[the remark this answers is not in the document]
Comment (smirnoal) [reply aws-lambda-java-tests/src/main/java/com/amazonaws/services/lambda/runtime/tests/EventLoader.java]: there are some compilation errors to resolve

[the remark this answers is not in the document]
Comment (mo7ty) [reply aws-lambda-java-tests/src/main/java/com/amazonaws/services/lambda/runtime/tests/EventLoader.java]: Think that they're solved with commit 83048d58.

**Review by andclt — APPROVED** [review]

**Comment by smirnoal** [comment]
Thanks for your contribution @mo7ty. The updated serialisation library has been published to Maven and can be consumed now: https://central.sonatype.com/artifact/com.amazonaws/aws-lambda-java-serialization

**Pull request merged**